lib/core/collections.toit

```diff
@@ -1078,6 +1078,16 @@ interface ByteArray:
   */
   do --reversed/bool [block] -> none
 
+  /**
+  Maps every byte with a mapping block.
+  The block is invoked with an integer argument in the range between 0 and
+    0xff, inclusive.
+  If all invocations of the block return integers between 0 and 0xff,
+    inclusive, then the return value is a byte array.
+  If some invocations return other values, the return value is a List.
+  */
+  map [block] -> any
```

Comment:
Not sure I like the dynamic changing.
Maybe:
```
map --bytes/bool [block] -> ByteArray:
  if not bytes: throw "INVALID ARGUMENT"

map [block] -> List
```

Edit: I see... later this is used for unicode. Not sure how to handle that.

Reply:
Hmm.
what about:

```
/** Maps bytes to bytes. */
map --in_place/bool=false --bytes/bool [block] -> ByteArray

/** Maps UTF8 characters to UTF8 characters */
map --unicode/bool [block] -> ByteArray
// We could even consider:
/**
Maps UTF8 characters to UTF8 characters.
If $in_place is true, attempts to store the result in this byte array.
  If the resulting byte-array is not the same size as this byte array allocates a fresh one and returns it.
  This byte-array should not be used anymore.
*/
map --in_place/bool=false --unicode/bool -> ByteArray

/** Maps each byte to the result of calling the given $block. */
map [block] -> List
```

Reply:
Admittedly, the wording "is not the same size" is not enough (as a single-byte character could explose to multiple bytes, before later characters shrink the size of the byte array again).

Reply:
The ByteArray code does not handle Unicode in any way.  You are thinking about the string code.

Reply:
I'm suggesting that it should know about it. The main reason that you want to be able to return either an int or a byte is for unicode.

Reply:
You can always do to_string.map if you want to handle Unicode. 

A byte array can contain byte sequences that are not valid UTF-8.  Feels wrong to mix Unicode into byte array operations. 

The main reason to allow return values that are not ints in the 0-255 range is if you were using the byte array as a more memory compact list of ints, but you don't want to be similarly restricted after map.

eg `(my_byte_array.map: "$(%02x it)").join ""`  // Convert #[1, 2, 3] to "010203"

Reply:
I don't think we should optimize our APIs for such a use-case. If you have a list of numbers, then you should use a list; potentially wrapping a byte-array for more efficient storage.  ByteArrays and Lists are obviously related, but there is a reason they don't share an interface or common super class. One is data, the other is sequence of integers. I don't want to have a type (-> any) where we mix them up. That's the reason I suggested to have different functions.